java-checks/src/main/java/org/sonar/java/checks/CollectionMethodsWithLinearComplexityCheck.java

```diff
@@ -129,14 +138,34 @@ private static Symbol invocationTarget(MethodInvocationTree mit) {
     return null;
   }
 
-  private static Type actualType(Symbol symbol) {
+  private static Set<String> findAssignedTypes(Symbol symbol) {
+    Set<String> types = new HashSet<>();
     Tree declaration = symbol.declaration();
     if (declaration != null && declaration.is(Tree.Kind.VARIABLE)) {
       ExpressionTree initializer = ((VariableTree) declaration).initializer();
       if (initializer != null) {
-        return initializer.symbolType();
+        types.add(initializer.symbolType().fullyQualifiedName());
       }
     }
-    return symbol.type();
+    symbol.usages().stream()
+      .flatMap(CollectionMethodsWithLinearComplexityCheck::usageInAssignment)
+      .map(assignment -> assignment.expression().symbolType().fullyQualifiedName())
+      .forEach(types::add);
+    return types;
+  }
+
+  private static Stream<AssignmentExpressionTree> usageInAssignment(IdentifierTree usage) {
+    Tree parent = usage.parent();
+    while (parent != null && !parent.is(Tree.Kind.ASSIGNMENT)) {
+      parent = parent.parent();
+    }
+    if (parent == null) {
+      return Stream.empty();
+    }
+    AssignmentExpressionTree assignment = (AssignmentExpressionTree) parent;
+    if (ExpressionUtils.skipParentheses(assignment.variable()).equals(usage)) {
```

Comment:
This won't work if assignment is done with member access: `this.myField = new CopyOnWriteArrayList<String>();`

You have to check that while going through parents, you are coming from the `variable()` side of the assignment, and not the `expression()` side. Checking only for `IdentifierTree` is not precise enough.